servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/LingeringRoundRobinLoadBalancerTest.java

```diff
@@ -0,0 +1,166 @@
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        final Predicate<TestLoadBalancedConnection> connectionFilter = alwaysNewConnectionFilter();
+
+        TestLoadBalancedConnection conn1 = lb.selectConnection(connectionFilter).toFuture().get();
+        TestLoadBalancedConnection conn2 = lb.selectConnection(connectionFilter).toFuture().get();
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 2));
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 2));
+
+        conn1.closeAsync().toFuture().get();
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 1));
+
+        conn2.closeAsync().toFuture().get();
+        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
+    }
+
+    @Test
+    public void closedConnectionDoesntRemoveActiveHost() throws Exception {
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        final Predicate<TestLoadBalancedConnection> connectionFilter = alwaysNewConnectionFilter();
+
+        TestLoadBalancedConnection conn1 = lb.selectConnection(connectionFilter).toFuture().get();
+        TestLoadBalancedConnection conn2 = lb.selectConnection(connectionFilter).toFuture().get();
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 2));
+        conn1.closeAsync().toFuture().get();
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 1));
+        conn2.closeAsync().toFuture().get();
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 0));
+    }
+
+    @Test
+    public void handleDiscoveryEventsForConnectedHosts() throws Exception {
+        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
+        assertThat(lb.activeAddresses(), is(empty()));
+
+        final Predicate<TestLoadBalancedConnection> connectionFilter = alwaysNewConnectionFilter();
+
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+        // For an added host, connection needs to be initiated, otherwise the host is free to be deleted
+        lb.selectConnection(connectionFilter).toFuture().get();
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+
+        sendServiceDiscoveryEvents(upEvent("address-2"));
+        lb.selectConnection(connectionFilter).toFuture().get();
+
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+
+        sendServiceDiscoveryEvents(downEvent("address-3"));
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        // At this moment in time, duplicates are allowed and the down event removes just the first address
+        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
```

Comment:
Sure, the duplicates are allowed at this moment. However, there is a difference between seeing two available events in a row for the same address and going back and forth between states ("available" -> "unavailable" -> "available") for the same address.

`DefaultServiceDiscoveryRetryStrategy` guarantees that it won't propagate duplicates, but the status change is not considered as a duplicate. Next "available" event should clear `isExpired` flag and make this connection availalbe again. Only if an address is already "availalbe", the next "available" event can add a new `Host` entry. Same for two "unavailable" events in a row:
```
upEvent("address-1")
// open connection
upEvent("address-1")
// open connection
downEvent("address-1")
downEvent("address-1")
``` 
^ in this case both have to be marked as "expired". The next `upEvent("address-1")` will clear `isExpired` flag from the first host only.

Reply:
That's a great point. I'm shocked to have missed it. The next commit covers this case and simplifies the branching logic of handling SD events too.